### PR TITLE
Create modulescanner.csv while walking the target

### PR DESCRIPTION
**Pull request by sormuras**
Produces `modulescanner.csv` file:

```csv
groupId,artifactId,version,moduleName,moduleVersion,moduleMode,moduleDependencies,jdepsToolError,jdepsViolations
com.fasterxml.jackson.core,jackson-core,2.9.6,com.fasterxml.jackson.core,-,automatic,-,false,-
org.slf4j,slf4j-api,1.8.0-beta2,org.slf4j,-,explicit,java.base,false,-
```

### CODE REVIEW

File: src/main/java/org/adoptopenjdk/modulescanner/Main.java [hunk 1]
```diff
@@ -19,25 +26,42 @@
...
      *
      * @param args Commandline arguments
      */
-    public static void main(String[] args) {
+    public static void main(String... args) {
         var directoryToScan = Paths.get(args.length > 0 ? args[0] : DEFAULT_DIRECTORY_TO_SCAN);
         var cutoffDate = args.length > 1 ? args[1] : CUTOFF_DATE;
+        var outputFileName = args.length > 2 ? args[2] : DEFAULT_OUTPUT_FILE_NAME;
 
+        // walk(directoryToScan, cutoffDate, new PrintWriter(System.out));
```
[the remark this answers is not in the document]
Comment (sormuras) [reply src/main/java/org/adoptopenjdk/modulescanner/Main.java]: Removed.

File: src/main/java/org/adoptopenjdk/modulescanner/Main.java [hunk 2]
```diff
@@ -19,25 +26,42 @@
...
+
+        try (var fw = new FileWriter(outputFileName); var bw = new BufferedWriter(fw); var out = new PrintWriter(bw)) {
+            walk(directoryToScan, cutoffDate, out);
+        } catch (IOException ioe) {
+            LOGGER.error("Walking directory: " + directoryToScan + " failed", ioe);
+            throw new UncheckedIOException("", ioe);
```
[the remark this answers is not in the document]
Comment (sormuras) [reply src/main/java/org/adoptopenjdk/modulescanner/Main.java]: Does it make sense to continue when an `IOException` occurs during creating the output file?

Removed the `throw new UncheckedIOException` line.

File: src/main/java/org/adoptopenjdk/modulescanner/Main.java [hunk 3]
```diff
@@ -50,4 +74,34 @@ private static JarFile toJarFile(Path path) {
         }
     }
 
+    // Create CSV header as string
```
[the remark this answers is not in the document]
Comment (sormuras) [reply src/main/java/org/adoptopenjdk/modulescanner/Main.java]: Sure. Makes sense.

**Comment by sormuras** [comment]
Refactored according to your suggestions, @karianna

File: src/test/java/org/adoptopenjdk/modulescanner/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import org.junit.jupiter.api.Test;
+
+class MainTest {
+
+    @Test
+    void testCsvOutput() throws Exception {
+        var testMavenRepo = Paths.get("src/test/resources/test-maven-repo");
+        var targetOutput = Paths.get("target/actual-modulescanner.csv");
```
Comment (sandermak) [inline src/test/java/org/adoptopenjdk/modulescanner/MainTest.java]: Is this guaranteed to be cleaned up between runs? Normally I'd use https://docs.oracle.com/javase/10/docs/api/java/io/File.html#createTempFile(java.lang.String,java.lang.String)

File: src/main/java/org/adoptopenjdk/modulescanner/Main.java [hunk 5]
```diff
@@ -19,25 +23,48 @@
...
+            walk(directoryToScan, cutoffDate, out);
+            LOGGER.info("Wrote " + Files.size(output) + " bytes to: " + output);
+        } catch (IOException ioe) {
+            LOGGER.error("Creating output file " + output + " failed", ioe);
+        }
+    }
+
+    // Walk repository and emit CSV file as output
+    private static void walk(Path directoryToScan, String cutoffDate, PrintWriter out) {
+        var csv = new CsvPrinter(out, ", ");
```
Comment (sandermak) [inline src/main/java/org/adoptopenjdk/modulescanner/Main.java]: Spaces are significant in csv, so better leave the trailing space off.

File: src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+				valueOrDashIfBlank(mir.moduleName),
+				valueOrDashIfBlank(mir.moduleVersion),
+				mir.isAutomaticModule ? "automatic" : mir.isExplicitModule ? "explicit" : "-",
+				valueOrDashIfBlank(String.join(" + ", mir.dependencies)));
+		return String.join(delimiter, columns);
+	}
+
+	// Convert null or blank values to "-".
+	private static String valueOrDashIfBlank(String value) {
+		return (value == null || value.trim().isEmpty() ) ? "-" : value;
```
Comment (sandermak) [inline src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java]: Not sure if the dash will help or hurt, but that's easily changed later.

File: src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	// Note: Keep in sync with #generateLine
+	private String generateHeaderLine() {
+		var columns = List.of(
+				"groupId",
+				"artifactId",
+				"version",
+				"moduleName",
+				"moduleVersion",
+				"moduleMode",
+				"moduleDependencies");
```
Comment (sandermak) [inline src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java]: Maybe also include jdeps violation count?

File: src/test/java/org/adoptopenjdk/modulescanner/MainTest.java (hunk 4 above)
Comment (sormuras) [reply src/test/java/org/adoptopenjdk/modulescanner/MainTest.java]: It is replaced by each run. But using a random and temporay target is better here. 👍

File: src/main/java/org/adoptopenjdk/modulescanner/Main.java (hunk 5 above)
Comment (sormuras) [reply src/main/java/org/adoptopenjdk/modulescanner/Main.java]: Sure. Or switch to tabs as delimiter? Without spaces `"\t"` and renaming the helper class to `SeparatedValuesPrinter`?

File: src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java (hunk 6 above)
Comment (sormuras) [reply src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java]: Not sure either. At least it is consistent.

File: src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java (hunk 7 above)
Comment (sormuras) [reply src/main/java/org/adoptopenjdk/modulescanner/CsvPrinter.java]: Aye. What else?

**Pull request merged**